### PR TITLE
update Sigstore community roadmap

### PR DESCRIPTION
**Pull request by bobcallaway**
Hi everyone!

The @sigstore/core-team has been working on updating the community's roadmap through conversations with our maintainers, users, and others in the wider OSS ecosystem. Our draft is now at a point where we'd like to solicit broader community feedback.

We are targeting to merge this by September 15th, 2023; however, our intent is that this is a living document, so PRs are always welcome :)

### CODE REVIEW

**Comment by oej** [comment]
The time stamp server seems missing in this document.

**Comment by haydentherapper** [comment]
@oej, under the transparency log roadmap, we mention "Timestamp at Entry Upload", so that TSA timestamps are included in the log and become immutable, and "Roughtime for distributed trusted timestamps", which is a long-term vision for distributed trusted timestamping. The timestamp server at https://github.com/sigstore/timestamp-authority is feature complete at this point. Was there anything you thought we needed to call out?

**Comment by oej** [comment]
@haydentherapper I just noted that it was missing and wondered if it was a mistake. Obviously not. Maybe add a comment that the time stamp server is feature complete if others wonder too :-)

Thank you for the feedback!

File: ROADMAP.md [hunk 1]
```diff
@@ -1,25 +1,389 @@
...
+
+## Current Status (as of September 2023)
+
+At the time of writing, the public Rekor signature transparency log contains over 32 million entries representing signatures from more than 17,000 unique open source projects including Kubernetes, CPython, Istio, urllib3, and Tekton (more [adopters](https://github.com/sigstore/friends) can be seen at the Sigstore [project landscape](https://landscape.openssf.org/sigstore)). Sigstore’s certificate authority has issued more than 21 million short-lived code signing certificates based on the OpenID Connect identity binding approach.  
+
+Sigstore announced the [general availability](https://openssf.org/press-release/2022/10/25/sigstore-announces-general-availability-at-sigstorecon/) of its public-good services in October of 2022 which built upon the v1.0 releases of Cosign, Fulcio, and Rekor. The public services have consistently met their 99.5% availability SLO since that milestone.
+
+Sigstore has several implementations of its core signing and verification workflows for popular programming languages, including:
+
+| Programming Language | Repository               |       Status      |
```
Comment (loosebazooka) [inline ROADMAP.md:44]: anything for sigstore-js?

File: ROADMAP.md [hunk 2]
```diff
@@ -1,25 +1,389 @@
...
+* [Cosign CLI](https://github.com/sigstore/cosign) is the official Sigstore CLI multi-tool for signing and verification in a variety of environments and use cases.
+* [sigstore-policy-controller](https://github.com/sigstore/policy-controller) is the center of Sigstore-oriented verification flows in Kubernetes environments.
+
+# Roadmap
+
+## Client SDKs
+
+### Short Term (< 6 mo)
+
+* Releases:
```
Comment (loosebazooka) [inline ROADMAP.md:204]: anything for sigstore-js here?

File: ROADMAP.md [hunk 3]
```diff
@@ -1,25 +1,389 @@
...
+* Privacy-preserving identities
+    * VRFs or ZKP to prevent attacker from learning when a signing event occurred for a given identity
+    * Must still support identity monitoring
+* Zero-knowledge proofs that prove a CA verified an identity token
+
+### Long Term (12+ mo)
+
+* ACME for code-signing certificates (ongoing RFC)
+* Native SPIFFE support with X509-SVIDs as authenticators, rather than supporting only OIDC
+* Support revocation for private deployments
```
Comment (mnm678) [inline ROADMAP.md:311]: is revocation for the public deployment intentionally omitted? This could be useful for key or account compromise (discussed in #sigstore/cosign#677)

Comment (haydentherapper) [reply ROADMAP.md:311]: It's a hard problem on a large scale - How do we authenticate requests to revoke content? How do we maintain an indefinitely growing list? WebPKI doesn't deal with revocation well, and I worry we'd run into the same issues.

File: ROADMAP.md [hunk 4]
```diff
@@ -1,25 +1,389 @@
...
+* TUF root supports additional delegations
+
+### Medium Term (6-12 mo)
+
+* BigQuery dataset for researchers
+* Alignment within OpenSSF (SCI WG, Security Toolbelt)
+
+### Long Term (12+ mo)
+
+* Sigstore API specified in IETF RFC
```
Comment (haydentherapper) [inline ROADMAP.md:389]:
```suggestion
* Sigstore API specified in IETF RFC

## Timestamp Authority

Feature complete. Future improvements include integrating timestamps into Rekor entries and exploring Roughtime as an alternative for distributed trusted timestamping.
```

File: ROADMAP.md (hunk 3 above)
Comment (mnm678) [reply ROADMAP.md:311]: I agree that this is challenging, and should be part of the 12 months+ roadmap. But if Rekor is used to discover a mis-behaving key/identity, Sigstore should do *something* about this. Whether this is a revocation list, TUF delegations to trusted identities, or some other mechanism.

Comment (jalseth) [reply ROADMAP.md:311]: Is this revoking specific entries in the log by UUID, revoking everything signed with public key X, revoking everything signed by x509 cert with serial Y, or revoking everything signed by `foo@bar.com` between a date range?

**Comment by znewman01** [comment]
Discussed in SIG-clients meeting today. Overall looks good to us + compatible with our roadmap, we'll leave specific comments as warranted.

File: ROADMAP.md (hunk 3 above)
Comment (jalseth) [reply ROADMAP.md:311]: I that case I agree. I think it is up to the consumers to determine what is trustworthy. 

I imagine package manager services would manage this in most cases so it would be transparent to the user, and the operators of each package service can handle authz for determining who can revoke trust.

Comment (mnm678) [reply ROADMAP.md:311]: I agree that the policy should be up to consumers, but Sigstore can provide tooling to make these revocations possible

Comment (trishankatdatadog) [reply ROADMAP.md:311]: @mnm678 Could the PKI for public deployments (including policies) be covered by [Line 380](https://github.com/sigstore/community/pull/315/files#diff-683343bdf93f55ed3cada86151abb8051282e1936e58d4e0a04beca95dff6e51R380)?

Comment (mnm678) [reply ROADMAP.md:311]: possibly, depending on how that is implemented

Comment (haydentherapper) [reply ROADMAP.md:311]: Do you have suggestions of wording changes to either this bullet point or the one Trishank noted that would cover what you're looking for? A mention of "client support for revocation policies"?

Comment (mnm678) [reply ROADMAP.md:311]: maybe "tooling to support client revocation policies"

Comment (trishankatdatadog) [reply ROADMAP.md:311]: I'd say it's more general than just revocation, but something along the lines of: "A solution for distributing, revoking, and rotating public keys and policies used to verify public artifacts"

Comment (haydentherapper) [reply ROADMAP.md:311]: I think that fits under the client roadmap with "Policy: sophisticated verification policies, or deeper integration with a policy evaluation service".

File: ROADMAP.md [hunk 5]
```diff
@@ -1,25 +1,389 @@
...
+* Privacy-preserving telemetry (https://divviup.org/)
+
+### Long Term (12+ mo)
+
+* Plumbing and porcelain model for Cosign, to improve Cosign's user experience
+* Policy: sophisticated verification policies, or deeper integration with a policy evaluation service
+* Further integration with TUF through delegations or trusted verification material distribution, and in-toto for signed provenance
+* Adding additional conformance tests to verify consistent behavior across clients, including negative tests
+* Bundle format includes attestation support that interops with SLSA
+* Reference implementation of how to use signature verification across various points in the software development lifecycle, leveraging policy-controller as one example
```
Comment (haydentherapper) [inline ROADMAP.md:244]:
```suggestion
* Reference implementation of how to use signature verification across various points in the software development lifecycle, leveraging policy-controller as one example
* Tooling to support client revocation policies
```

File: ROADMAP.md [hunk 6]
```diff
@@ -1,25 +1,389 @@
...
+    * Monitors will only verify log consistency
+* Witnessing
+    * Monitors should transition into fully fledged witnesses
+    * Witnesses should publish signed checkpoints to distributor
+* V1 for GitHub Actions-based identity monitor
+* Timestamp at Entry Upload
+    * Timestamp should be provided by TSA
+    * Timestamp becomes immutable and verifiable
+* Complete architecture document
+* Publish inclusion events to pub/sub queueo
```
Comment (haydentherapper) [inline ROADMAP.md:260]:
```suggestion
* Publish inclusion events to pub/sub queue
```

File: ROADMAP.md (hunk 2 above)
Comment (haydentherapper) [reply ROADMAP.md:204]: @bdehamer Anything you wanted to call out? Any major releases in the next year?

File: ROADMAP.md [hunk 7]
```diff
@@ -1,25 +1,389 @@
...
+* Complete architecture document, get commitment on operational transfer
+
+### Medium Term (6-12 mo)
+
+* Begin transition of public instance to neutral third party
+* Release automation
+    * New releases should go to staging automatically
+    * Releases should seamlessly move from staging to production with automated testing
+* New bleeding-edge environment
+    * Hourly/nightly releases should automatically deploy to this env
```
[the remark this answers is not in the document]
Comment (haydentherapper) [reply ROADMAP.md:365]: That's a good callout, we have an issue in the private infra repo for tracking automatically filing issues on GH, but this has not been completed.

File: ROADMAP.md [hunk 8]
```diff
@@ -1,25 +1,389 @@
...
+* Availability improvements
+    * Regionalization of components
+    * Read replicas of log
```
Comment (haydentherapper) [inline ROADMAP.md:368]:
```suggestion
    * Read replicas of log
* Improve automated issue filing for alerts
```

File: ROADMAP.md [hunk 9]
```diff
@@ -1,25 +1,389 @@
...
+
+### Long Term (12+ mo)
+
+* Public root of trust trusted by private ecosystems (Apple App Store, Google Play Store, Windows)
```
[the remark this answers is not in the document]
Comment (haydentherapper) [reply ROADMAP.md:372]: Policy would need to be a part of this, yes.

File: ROADMAP.md (hunk 3 above)
Comment (oej) [reply ROADMAP.md:311]: I suggest we don't use the word "revocation" until we had a discussion about it. Avoiding any "revocation" solution is generally a good thing, since most of them has not worked very well. So Hayden's suggestion seems good.

Comment (trishankatdatadog) [reply ROADMAP.md:311]: Right, that makes sense: each, say, package registry may wish to handle revocation separately from Sigstore. Sigstore should ideally "just" be a (but not the only) source of truth.

File: ROADMAP.md [hunk 10]
```diff
@@ -1,25 +1,389 @@
...
+
+Sigstore announced the [general availability](https://openssf.org/press-release/2022/10/25/sigstore-announces-general-availability-at-sigstorecon/) of its public-good services in October of 2022 which built upon the v1.0 releases of Cosign, Fulcio, and Rekor. The public services have consistently met their 99.5% availability SLO since that milestone.
+
+Sigstore has several implementations of its core signing and verification workflows for popular programming languages, including:
+
+| Programming Language | Repository               |       Status      |
+|:--------------------:|--------------------------|:-----------------:|
+|        Golang        | [sigstore/cosign](https://github.com/sigstore/cosign)          |         GA        |
```
Comment (trevrosen) [inline ROADMAP.md:46]: Imo it's best to refer to the Go programming language as "Go" and leave "Golang" for use as a search term. 😄

File: ROADMAP.md [hunk 11]
```diff
@@ -1,25 +1,389 @@
...
+* Further integration with TUF through delegations or trusted verification material distribution, and in-toto for signed provenance
+* Adding additional conformance tests to verify consistent behavior across clients, including negative tests
+* Bundle format includes attestation support that interops with SLSA
+* Reference implementation of how to use signature verification across various points in the software development lifecycle, leveraging policy-controller as one example
+
+## Signature Transparency Log (Rekor)
+
+### Short Term (< 6 mo)
+
+* Add 2 independent log monitors
```
Comment (trevrosen) [inline ROADMAP.md:250]: Should we call out what we want to use to monitor the logs? I know we have a monitoring project here in the Sigstore org that I believe powers the Purdue monitor. Should there be a roadmap for that kind of functionality?

Comment (haydentherapper) [reply ROADMAP.md:250]: Good suggestion, I’ll add a section on that. I have a roadmap for what’s needed to enable witnessing and identity monitoring.

File: ROADMAP.md (hunk 2 above)
Comment (bdehamer) [reply ROADMAP.md:204]: No major releases planned. I've got a long list of small features/improvements I'll be working on but nothing worth calling-out here.

File: ROADMAP.md (hunk 10 above)
Comment (haydentherapper) [inline ROADMAP.md:46]:
```suggestion
|        Go        | [sigstore/cosign](https://github.com/sigstore/cosign)          |         GA        |
```

File: ROADMAP.md [hunk 12]
```diff
@@ -1,25 +1,389 @@
...
+Sigstore has several implementations of its core signing and verification workflows for popular programming languages, including:
+
+| Programming Language | Repository               |       Status      |
+|:--------------------:|--------------------------|:-----------------:|
+|        Golang        | [sigstore/cosign](https://github.com/sigstore/cosign)          |         GA        |
+|        Python        | [sigstore/sigstore-python](https://github.com/sigstore/sigstore-python) |         GA        |
+|      Javascript      | [sigstore/sigstore-js](https://github.com/sigstore/sigstore-js)     |         GA        |
+|         Java         | [sigstore/sigstore-java](https://github.com/sigstore/sigstore-java)   |        Beta       |
+|         Rust         | [sigstore/sigstore-rs](https://github.com/sigstore/sigstore-rs)     |        Beta       |
+|        Golang        | [sigstore/sigstore-go](https://github.com/sigstore/sigstore-go)     | Under development |
```
Comment (haydentherapper) [inline ROADMAP.md:51]:
```suggestion
|        Go            | [sigstore/sigstore-go](https://github.com/sigstore/sigstore-go)     | Under development |
```

File: ROADMAP.md [hunk 13]
```diff
@@ -1,25 +1,389 @@
...
+
+## Brief Overview of Sigstore
+
+[Sigstore](https://sigstore.dev) was founded in June of 2020 to address the fundamental challenges preventing wide scale usage of digital signatures within the software supply chain. Sigstore is a set of open source projects and services that dramatically simplify the creation and verification of digital signatures. It achieves this through a modular architecture that integrates proven technology solutions from adjacent domains (including Certificate Transparency, OpenID Connect, and Workload Identity among others) while reusing industry standard cryptographic algorithms and best practices. 
+
+Sigstore is inherently flexible and can natively support a variety of:
+
+* deployment scenarios (e.g. public-facing VS private enterprise deployments), 
+* key management approaches (e.g. hardware-backed keys stored in HSMs or Yubikeys or ephemeral scenarios where a private key can be deleted immediately after use),
```
[the remark this answers is not in the document]
Comment (haydentherapper) [reply ROADMAP.md:27]: Deleted could also be cleared from memory.

File: ROADMAP.md [hunk 14]
```diff
@@ -1,25 +1,389 @@
...
+* Verifiable Search Implementation
+* Package repositories participate in mitigating split-views
+    * Witnessed checkpoints attached to artifact signature metadata
+    * Repositories verify log consistency
+* Implementation for log replication
+* Transparent redaction implementation
+* Privacy-preserving log queries, where the log is not aware of client's artifact of interest
+* Roughtime for distributed trusted timestamps
+* Continued fuzzing of large API surface
+
```
Comment (haydentherapper) [inline ROADMAP.md:289]:
```suggestion

## Transparency Log Monitor

### Short Term (< 6 mo)

* Initial launch of Sigstore's [log monitor on GitHub Actions](https://github.com/sigstore/rekor-monitor)
   * Monitors acts as both a witness (co-signing checkpoints) and Verifier (monitoring given identities)
   * Co-signed checkpoints are published to GitHub
   * Verifier monitors for OIDC identities and key fingerprints
   * Alerts are filed as GitHub issues
   * Supports sharded log
* At least 2 independent instances of GitHub Actions-based log monitor running and publicly documented
* Publish blog posts on monitoring requirements as an artifact producer
* Create log cloning tool to support offline log monitoring and analysis
   * Can also be used to transform an existing log into a serverless layout

### Medium Term (6-12 mo)

* V1 launch of Sigstore's [log monitor on GitHub Actions](https://github.com/sigstore/rekor-monitor)
   * Witnessed checkpoints are published to distributors, operated by Sigstore and/or community members
   * Support monitoring Fulcio's CT log
   * Alerts are sent in a variety of ways (e.g. email)
   * Productionized and resilient against intermittent failures
* Publish select monitors to status.sigstore.dev

### Long Term (12+ mo)

* Optionality around running witness in different environments, e.g. hardware or cloud
* Integration with other efforts in transparency ecosystems
   * Publishing checkpoints to external community-operated distributors
* Collaborate with CT log monitors (e.g. [sslmate](https://sslmate.com/certspotter/)) to improve identity monitoring

```

Comment (haydentherapper) [reply ROADMAP.md:289]: @trevrosen PTAL and let me know if you see anything that's missing.

File: ROADMAP.md (hunk 11 above)
Comment (haydentherapper) [reply ROADMAP.md:250]: Added roadmap, PTAL!

File: ROADMAP.md [hunk 15]
```diff
@@ -1,25 +1,389 @@
...
+### Short Term (< 6 mo)
+
+* Publish document on public vs private deployments
+* Maintain operational excellence:
+    * 99.5% SLO
+    * Multi-vendor on call rotation
+    * Continued burn down of open tickets
+    * Periodic DiRT testing
+* Perform 1 log sharding and log key rotation
+* Complete architecture document, get commitment on operational transfer
```
Comment (haydentherapper) [inline ROADMAP.md:356]:
```suggestion
* Complete architecture document, get commitment on operational transfer
* Migrate TUF root repository over to TUF-on-CI
```

File: ROADMAP.md [hunk 16]
```diff
@@ -1,25 +1,389 @@
...
+
+### Long Term (12+ mo)
+
+* Terraform for non-GCP architecture
+
+## Documentation
+
+### Short Term (< 6 mo)
+
+* Documentation restructured around user journeys
```
Comment (haydentherapper) [inline ROADMAP.md:333]:
```suggestion
* Documentation restructured around user journeys
* Document workload identity/automated signing in addition to developer-driven signing
* Terminology page
```

File: ROADMAP.md (hunk 14 above)
Comment (trevrosen) [reply ROADMAP.md:290]: Looks great - thanks!

File: ROADMAP.md [hunk 17]
```diff
@@ -1,25 +1,429 @@
...
+5. Once RFC proposal is approved per community norms, drive implementation plan
+6. Work to drive initial package adoption strategy; this often involves highly visible or valuable packages that can serve as references for other package maintainers within the ecosystem
+7. Work with popular build & packaging extensions (e.g. GoReleaser, JReleaser) and builder templates (Jenkins plugins, GitHub Actions, 
+8. Release all code & service extensions required for Sigstore support in a fully-supported mode
+9. Presuming success to this point, set date for mandating Sigstore signing and verification for all packages in ecosystem 
+
+## Current Ecosystem Adoption Status
+
+The npm registry, the world's largest open source package manager for the Node.js ecosystem, [recently announced built-in support for Sigstore-signed SLSA build provenance](https://github.blog/2023-04-19-introducing-npm-package-provenance/) to users. This allows consumers of packages from npm to independently verify the link between a package fetched from the npm registry to the source code from which it was built. The build provenance attestation is cryptographically signed using Sigstore.
```
Comment (steiza) [inline ROADMAP.md:72]:
```suggestion
The npm registry, the world's largest open source package manager for the JavaScript ecosystem, [recently announced built-in support for Sigstore-signed SLSA build provenance](https://github.blog/2023-04-19-introducing-npm-package-provenance/) to users. This allows consumers of packages from npm to independently verify the link between a package fetched from the npm registry to the source code from which it was built. The build provenance attestation is cryptographically signed using Sigstore. Other package managers may find the [documentation of npm's Sigstore adoption and resulting design decisions](https://repos.openssf.org/build-provenance-for-all-package-registries) useful.
```

**Review by SantiagoTorres — APPROVED** [review]

**Review by trevrosen — APPROVED** [review]

**Review by cpanato — APPROVED** [review]

File: ROADMAP.md (hunk 3 above)
Comment (trishankatdatadog) [reply ROADMAP.md:343]: Do we have a resolution about this? @mnm678 @haydentherapper

Comment (mnm678) [reply ROADMAP.md:343]: Hayden's interpretation of it fitting under the policy item works for me

**Comment by bobcallaway** [comment]
Thanks to everyone for their feedback - again the intent is for this to be a living document, so please feel free to submit a PR to update this at anytime.
